Client/Frontend/Widgets/BottomSheetViewController.swift

```diff
@@ -0,0 +1,248 @@
+                yPosition = initialY
+            } else if yPosition > maxY { // Dragged all way down, remove the view.
+                yPosition = maxY
+            }
+            
+            update()
+        }
+    }
+    
+    private var initialDrawerYPosition: CGFloat {
+        let popupY = (view.frame.height / 2) - (contentView.frame.height / 2)
+        
+        let regularY = view.frame.maxY - contentView.frame.height
+        
+        return showAsPopup ? popupY : regularY
+    }
+    
+    private var showAsPopup: Bool {
+        traitCollection.userInterfaceIdiom == .pad && traitCollection.horizontalSizeClass == .regular
+    }
+    
+    // MARK: - Lifecycle
+
+    init() {
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    @available(*, unavailable)
+    required init?(coder aDecoder: NSCoder) { fatalError() }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        view.addSubview(backgroundOverlayView)
+        view.addSubview(contentView)
+        
+        contentView.backgroundColor = .white
+
+        let panGestureRecognizer = UIPanGestureRecognizer(target: self, action: #selector(didRecognizePanGesture))
+        contentView.addGestureRecognizer(panGestureRecognizer)
+
+        contentView.addSubview(handleView)
+        
+        contentView.addSubview(closeButton)
+        
+        closeButton.addTarget(self, action: #selector(closeView), for: .touchUpInside)
+        contentView.isHidden = true
+        
+        makeConstraints()
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        show()
+    }
+    
+    override func viewDidLayoutSubviews() {
+        yPosition = contentView.isHidden ? view.frame.maxY : initialDrawerYPosition
+        
+        updateDrawerViewConstraints()
+    }
+    
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
+        updateDrawerViewConstraints()
+    }
+    
+    // MARK: - Constraints setup
+    
+    private func makeConstraints() {
+        backgroundOverlayView.snp.makeConstraints { make in
+            let bottomInset = parent?.view.safeAreaInsets.bottom ?? 0
+            make.leading.trailing.top.equalToSuperview()
+            make.bottom.equalToSuperview().offset(bottomInset)
+        }
+        
+        handleView.snp.remakeConstraints {
+            $0.width.equalTo(UX.HandleWidth)
+            $0.height.equalTo(UX.HandleHeight)
+
+            $0.centerX.equalTo(contentView)
+            $0.top.equalTo(contentView).offset((UX.HandleMargin - UX.HandleHeight) / 2)
+        }
+        
+        closeButton.snp.makeConstraints {
+            $0.top.equalToSuperview().inset(10)
+            $0.right.equalToSuperview().inset(7)
+            $0.size.equalTo(26)
+        }
+        
+        updateDrawerViewConstraints()
+    }
+    
+    private func updateDrawerViewConstraints() {
```

Comment:
Why not just override `updateConstraints` and call `setNeedsUpdateConstraints()` in traitCollectionDidChange ?